packages/merkle-distributor/kpi-options-helpers/calculate-uma-tvl.ts

```diff
@@ -0,0 +1,154 @@
+    const priceNormalized = toBN(toWei(obj.collateralPriceInUsd));
+    const collateralValue = collateralBalanceNormalized.mul(priceNormalized).div(fixedPointAdjustment);
+
+    return { ...obj, collateralValueInUsd: fromWei(collateralValue) };
+  });
+}
+
+// For an array of collateral types associated with an array of financial contracts, compute the balance in collateral
+// of the financial contract, the collateral decimals and the value In USD of each unit of collateral.
+export async function fetchCollateralInfo(collateralAddresses: Array<string>, registeredContracts: Array<string>) {
+  const [collateralBalances, collateralDecimals, collateralPricesInUsd] = await Promise.all([
+    fetchCollateralBalances(collateralAddresses, registeredContracts),
+    fetchCollateralDecimals(collateralAddresses),
+    fetchCollateralValue(collateralAddresses)
+  ]);
+
+  return { collateralBalances, collateralDecimals, collateralPricesInUsd };
+}
+
+// Return an array of all registered contracts found in the Registry. Note that some non-financial contracts are included
+// in this list, such as the Optimistic oracle. These will have `null` values for their collateral type, balance and price.
+// They are filtered out later on in the `fetchAllFinancialContractsData` method.
+export async function fetchAllRegisteredContracts() {
+  const mainnetRegistry = new web3.eth.Contract(getAbi("Registry"), getAddress("Registry", 1));
+  const events = await mainnetRegistry.getPastEvents("NewContractRegistered", { fromBlock: 0, toBlock: "latest" });
+  return events.map((event: any) => event.returnValues.contractAddress);
+}
+
+// Returns an array of collateral types for an array of financial contract addresses. Note that `FeePayer` is the simplest
+// ABI implementation that provides the `collateralCurrency` public method.
+export async function fetchCollateralForFinancialContracts(financialContractAddresses: Array<string>) {
+  const contractInstances = financialContractAddresses.map(
+    (address: string) => new web3.eth.Contract(getAbi("FeePayer"), address)
+  );
+
+  // We use allSettled as some of the async calls might fail. In particular, if the method does not implement the public
+  // method collateralCurrency they will fail.
+  const collateralAddresses = await Promise.allSettled(
+    contractInstances.map(contractInstance => contractInstance.methods.collateralCurrency().call())
+  );
+
+  // If the contract does not have the method called then return null.
+  return collateralAddresses.map((response: any) =>
+    response.status === "fulfilled" ? response.value.toLowerCase() : null
+  );
+}
+
+// Returns an array of decimals associated with an array of collateral addresses. Note if any contract is not implemented
+// (null) or the decimals method is not implemented this will return null for that collateral token.
+export async function fetchCollateralDecimals(collateralTokenAddresses: Array<string>) {
```

Comment:
this is just fetching decimals from an erc20, doesnt need to be specific to collateral

Reply:
good point! it's not just for collateral. I renamed it to be contract decimals. I did the same thing for balances as well.